File: C++Verifier/src/expressions/BoolExpr.cpp
#ifndef BOOLEXPR_H
#define BOOLEXPR_H

#include "Expr.h"
#include "../DSFA/ConstraintTreeNode.cpp"

class BoolExpr: public virtual Expr {
    public:
        using Expr::Expr;

        virtual bool evaluate(shared_ptr<Packet> p,  vector<string> constraints, vector<string> locationList, vector<string> variableList, int currentState, vector<shared_ptr<ConstraintTreeNode>> rootConstraints) = 0;

        virtual void* evaluate(shared_ptr<Packet> p) = 0;

        void toggleNegated() {
            negated = !negated;
        }
        static bool applyBoolOp(string leftValue, string rightValue, string operand) {
            long leftVal = stol(leftValue);
            long rightVal = stol(rightValue);
            if (operand == ">") {
                return leftVal > rightVal;
            } else if (operand == ">=") {
                return leftVal >= rightVal;
            } else if (operand == "<") {
                return leftVal < rightVal;
            } else if (operand == "<=") {
                return leftVal <= rightVal;
            }
            // TODO: Throw exception heres

            return false;
        }
        virtual ~BoolExpr() {};
        bool negated = false;
};

#endif

File: verification/src/main/java/org/msr/mnr/verification/expressions/BoolExpr.java
package org.msr.mnr.verification.expressions;

import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;
import java.util.HashMap;

import org.msr.mnr.verification.dsfa.ConstraintTreeNode;
import org.msr.mnr.verification.utils.Packet;

public abstract class BoolExpr extends Expr {
    private static final long serialVersionUID = 2806048939951936622L;
    protected boolean negated = false;

    public BoolExpr(Object left, Object right) {
        super(left, right);
    }

    public BoolExpr() {
        super();
    }

    public BoolExpr(Object left, Object right, HashSet<String> locationReferences,
            HashMap<String, String> variableComparisons) {
        super(left, right, locationReferences, variableComparisons);
    }

    public void toggleNegated() {
        negated = !negated;
    }

    public abstract Boolean evaluate(Packet p, ArrayList<int[]> constraints,
            ArrayList<String> locationList, ArrayList<String> variableList, Integer currentState,
            List<ConstraintTreeNode> rootConstraints);

    @Override
    public abstract Boolean evaluate(Packet p);
}